src/sentry/monitors/clock_tasks/check_timeout.py

```diff
@@ -0,0 +1,99 @@
+from __future__ import annotations
+
+import logging
+from datetime import datetime
+
+from arroyo.backends.kafka import KafkaPayload
+from sentry_kafka_schemas.schema_types.monitors_clock_tasks_v1 import MarkTimeout
+
+from sentry.monitors.logic.mark_failed import mark_failed
+from sentry.monitors.models import CheckInStatus, MonitorCheckIn
+from sentry.monitors.schedule import get_prev_schedule
+from sentry.utils import metrics
+
+from .producer import MONITORS_CLOCK_TASKS_CODEC, produce_task
+
+logger = logging.getLogger("sentry")
+
+# This is the MAXIMUM number of pending MONITOR CHECKINS this job will check.
+#
+# NOTE: We should keep an eye on this as we have more and more usage of
+# monitors the larger the number of checkins to check will exist.
+CHECKINS_LIMIT = 10_000
+
+
+def dispatch_check_timeout(ts: datetime):
+    """
+    Given a clock tick timestamp determine which check-ins are past their
+    timeout_at.
+
+    This will dispatch MarkTimeout messages into monitors-clock-tasks.
+    """
+    qs = MonitorCheckIn.objects.filter(status=CheckInStatus.IN_PROGRESS, timeout_at__lte=ts)[
+        :CHECKINS_LIMIT
+    ]
+    metrics.gauge("sentry.monitors.tasks.check_timeout.count", qs.count(), sample_rate=0)
+    # check for any monitors which are still running and have exceeded their maximum runtime
+    for checkin in qs:
+        message: MarkTimeout = {
+            "type": "mark_timeout",
+            "ts": ts.timestamp(),
+            "checkin_id": checkin.id,
+        }
+        # XXX(epurkhiser): Partitioning by monitor_environment.id is important
+        # here as these task messages will be consumed in a multi-consumer
+        # setup. If we backlogged clock-ticks we may produce multiple timeout
+        # tasks for the same monitor_environment. These MUST happen in-order.
+        payload = KafkaPayload(
+            checkin.monitor_environment_id.to_bytes(),
+            MONITORS_CLOCK_TASKS_CODEC.encode(message),
+            [],
+        )
+        produce_task(payload)
+
+
+def mark_checkin_timeout(checkin_id: int, ts: datetime):
+    logger.info("checkin.timeout", extra={"checkin_id": checkin_id})
+
+    checkin = (
+        MonitorCheckIn.objects.select_related("monitor_environment")
+        .select_related("monitor_environment__monitor")
+        .get(id=checkin_id)
+    )
+
+    monitor_environment = checkin.monitor_environment
+    monitor = monitor_environment.monitor
+
+    logger.info(
+        "monitor_environment.checkin-timeout",
+        extra={"monitor_environment_id": monitor_environment.id, "checkin_id": checkin.id},
+    )
+    affected = checkin.update(status=CheckInStatus.TIMEOUT)
```

Comment:
This should be `MonitorCheckIn.objects.filter(id=checkin_id).exclude(status=CheckInStatus.TIMEOUT).update(status=CheckInStatus.TIMEOUT)`

Currently this will always return true for affected and makes the check basically useless

Reply:
Thanks for calling this out again, almost forgot

Reply:
Added a test for this here https://github.com/getsentry/sentry/pull/69928/files#diff-0e6aba69f0ecc20d43bc73796df8c1f6a0f0dd13668f547ee6fedca98e335525R618-R663

Reply:
NVM this comment was wrong. I will add a test.

Reply:
Correctly tested here now https://github.com/getsentry/sentry/pull/69928/files#diff-bd0465c5a31170959d7b7c9c2d8178630c1524bda85a4a17a6812ed29e81e383R443-R493